src/Microsoft.ML.Data/Evaluators/RankingEvaluator.cs

```diff
@@ -946,23 +948,34 @@ private protected override IEnumerable<string> GetPerInstanceColumnsToSave(RoleM
 
     internal static class RankingUtils
     {
+        private static readonly object _lock = new object();
         private static volatile Double[] _discountMap;
+        private static volatile int _maxTruncationLevel = 0;
+
         public static Double[] DiscountMap
         {
             get
             {
-                double[] result = _discountMap;
-                if (result == null)
+                return _discountMap;
+            }
+        }
+        /// <summary>
+        /// Reallocates discountMap for the largest truncationLevel seen so far
+        /// </summary>
+        public static void SetTruncationLevel(int truncationLevel)
```

Comment:
This shouldn't be `static`. There is something wrong with the existing code here. Using a global static `_discountMap` array seems bad. What happens if multiple threads are training/evaluating ranking models?

Reply:
The way that it's implemented now is by locking the `_discountMap` and resizing if the `truncationLevel` is larger then the `maxTruncationLevel`. This ensures that the `_discountMap` is always large enough. Let me know if there's a better approach.

Reply:
A better approach is to not use a static variable.

Reply:
RankingUtils is a static class so all its members will be static. I am guessing RankingUtils has been made static because the data in DiscountMap can be shared across threads once initialized. The implemented behavior right now will behave correctly even in a multi-threaded context.

---
In reply to: [327656304](https://github.com/dotnet/machinelearning/pull/4081#discussion_r327656304) [](ancestors = 327656304)